rasa_addons/core/nlg/bftemplate.py

```diff
@@ -85,11 +89,31 @@ async def generate(
             fallback_language=fallback_language,
         )
         if "language" in message: del message["language"]
+
+        if self.__is_image_available(message, "image"):
+            message["image"] = self.__substitute_url(message, "image", message["image"])
+        if self.__is_image_available(message, "elements"):
+            #loop over the image urls in a carousel
+            for element in message["elements"]:
+                element["image_url"] = self.__substitute_url(element, "image_url", element["image_url"])
+
         metadata = message.pop("metadata", {}) or {}
         for key in metadata: message[key] = metadata[key]
 
         return message
 
+    def __is_image_available(self, message: dict, value: str):
+        """Check if an image key is available in the message dict or image_url in carousel list."""
+        return value in message.keys() and self.url_substitution_pattern
+
+    def __substitute_url(self,message:dict, key: str, url:str):
```

Comment:
You don't need three arguments, just one, `url`, since all you do with the first two is access the third.

Reply:
We actually do need two arguments since the for images in carousel and single image is different. I removed the "url" since we can collect it by key.